### ISSUE TITLE
Optimize runtime history reindex

### ISSUE THREAD

**Issue by kostko** [oasisprotocol/oasis-core#5738]
Possibly doing the reindex in parallel. Should do some experiments.

**Comment by peternose** [comment]
@kostko Did you mean to reindex every runtime in parallel, or to reindex blocks or blocks of blocks in parallel?

**Comment by kostko** [comment]
Both if that would improve performance. We can split it into two issues though.

**Comment by ptrus** [comment]
:+1: for both. Some more context: Reindexing a single runtime already takes a lot of time, so we want/need to parallelize reindexing blocks in that scenario. This will likely have the biggest impact on speeding up the full node initialization. It probably also makes sense to parallelize it across runtimes as well.

One additional related thing that I remember from the past is that we currently block starting any runtimes until all have completed the history reindex. This could likely be improved as well. But probably belongs in a separate issue.

**Comment by martintomazic** [comment]
For the reference:

1. #6050 solved write bottleneck of history reindex (3x increase + enables read optimiziations).
2. #6095 stopped reindexing redundant  round results (further 30% increase).
3. #6089 moved history reindex from roothash service to runtime registry.
    * Runtimes are now automatically reindexed in parallel, so the case where reindex of one runtime blocks already reindexed runtimes is gone.


This gave us following performance gain:
1. `4x` speed-up for single runtime.
2. `4 x n` speed-up for `n` runtimes. (Likely if `n` grows this factor decreases).
 
E.g. having `cipher`, `emerald` and `sapphire` configured resulted in `12-14x` speed-up (benched on testnet starting with synced consensus and empty block history dir). We are happy with that for now, so closing this issue.

TODO:
We could easily get another 2-3x, by optimizing the reading part. This was already [benched](https://docs.google.com/document/d/1Lq-Qe6jrrLUSKIkLLsIGOrfb31Ivi-hZKP0wAC50V1E/edit?tab=t.0#heading=h.97gqwtgjc6xf), where we fetched reads in parallel within a batch (using worker pool - (#6055)).
